File: Text_Editor_Project/Core/History.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Text.Json;
using System.IO;

namespace Text_Editor_Project.Core
{
    public class History
    {
        public TextEditor textEditor;
        public List<EditorMemento> editorMementos;

        public History(TextEditor textEditor) {

            this.textEditor = textEditor;
        }

        public void Backup()
        {
            var editor= textEditor.CreateMemento();
            editorMementos.Add(editor);

            Console.WriteLine("Backup stored successfully.");



        }

        public void Undo()
        {

            if(editorMementos.Count > 0)
            {

                EditorMemento memento = editorMementos[editorMementos.Count-1];
                editorMementos.RemoveAt(editorMementos.Count-1);
                textEditor.RestoreFromMemento(memento);
                Console.WriteLine("Undo operation was successful.");
            }
            else
            {
                Console.WriteLine("nothing to restore.");
            }

        }

        public void SaveToFile(string filePath)
        {
            try
            {
                string json = JsonSerializer.Serialize(editorMementos, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filePath, json);
                Console.WriteLine("History saved successfully to file.");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error saving history: {ex.Message}");
            }


        }

        public void LoadFromFile(string filePath)
        {
            if (!File.Exists(filePath))
            {
                Console.WriteLine("File not found.");
                return;
            }
            string json = File.ReadAllText(filePath);
            editorMementos = JsonSerializer.Deserialize<List<EditorMemento>>(json) ?? new List<EditorMemento>();
            Console.WriteLine("History loaded successfully from file.");

        }


    }
}

File: TextEditorMemento/src/main/java/texteditor/History.java
package main.java.texteditor;

import java.util.Stack;

public class History {
    private final Stack<EditorMemento> historyStack = new Stack<>();
    private final TextEditor editor;

    public History(TextEditor editor) {
        this.editor = editor;
    }

    public void backup() {
        EditorMemento memento = editor.createMemento();
        historyStack.push(memento);
        System.out.println("Backup created.");
    }

    public void undo() {
        if (historyStack.isEmpty()) {
            System.out.println("No previous state to restore.");
            return;
        }
        EditorMemento memento = historyStack.pop();
        editor.restoreFromMemento(memento);
        System.out.println("Undo successful.");
    }
}
